user-contrib/Ltac2/g_ltac2.mlg

```diff
@@ -0,0 +1,933 @@
+        plugin = Prim.string; name = Prim.string ->
+      { let ml = { mltac_plugin = plugin; mltac_tactic = name } in
+        StrPrm (id, t, ml) }
+    ] ]
+  ;
+  syn_node:
+    [ [ "_" -> { CAst.make ~loc None }
+      | id = Prim.ident -> { CAst.make ~loc (Some id) }
+    ] ]
+  ;
+  sexpr:
+    [ [ s = Prim.string -> { SexprStr (CAst.make ~loc s) }
+      | n = Prim.integer -> { SexprInt (CAst.make ~loc n) }
+      | id = syn_node -> { SexprRec (loc, id, []) }
+      | id = syn_node; "("; tok = LIST1 sexpr SEP "," ; ")" ->
+        { SexprRec (loc, id, tok) }
+    ] ]
+  ;
+  syn_level:
+    [ [ -> { None }
+      | ":"; n = Prim.integer -> { Some n }
+    ] ]
+  ;
+  tac2def_syn:
+    [ [ "Notation"; toks = LIST1 sexpr; n = syn_level; ":=";
+        e = tac2expr ->
+        { StrSyn (toks, n, e) }
+    ] ]
+  ;
+  lident:
+    [ [ id = Prim.ident -> { CAst.make ~loc id } ] ]
+  ;
+  globref:
+    [ [ "&"; id = Prim.ident -> { CAst.make ~loc (QHypothesis id) }
+      | qid = Prim.qualid -> { CAst.make ~loc @@ QReference qid }
+    ] ]
+  ;
+END
+
+(* Quotation scopes used by notations *)
+
+{
+
+open Tac2entries.Pltac
+
+let loc_of_ne_list l = Loc.merge_opt (fst (List.hd l)) (fst (List.last l))
+
+}
+
+GRAMMAR EXTEND Gram
+  GLOBAL: q_ident q_bindings q_intropattern q_intropatterns q_induction_clause
+          q_conversion q_rewriting q_clause q_dispatch q_occurrences q_strategy_flag
+          q_destruction_arg q_reference q_with_bindings q_constr_matching
+          q_goal_matching q_hintdb q_move_location q_pose q_assert;
+  anti:
+    [ [ "$"; id = Prim.ident -> { QAnti (CAst.make ~loc id) } ] ]
+  ;
+  ident_or_anti:
+    [ [ id = lident -> { QExpr id }
+      | "$"; id = Prim.ident -> { QAnti (CAst.make ~loc id) }
+    ] ]
+  ;
+  lident:
+    [ [ id = Prim.ident -> { CAst.make ~loc id } ] ]
+  ;
+  lnatural:
+    [ [ n = Prim.natural -> { CAst.make ~loc n } ] ]
+  ;
+  q_ident:
+    [ [ id = ident_or_anti -> { id } ] ]
+  ;
+  qhyp:
+    [ [ x = anti -> { x }
+      | n = lnatural -> { QExpr (CAst.make ~loc @@ QAnonHyp n) }
+      | id = lident -> { QExpr (CAst.make ~loc @@ QNamedHyp id) }
+    ] ]
+  ;
+  simple_binding:
+    [ [ "("; h = qhyp; ":="; c = Constr.lconstr; ")" ->
+        { CAst.make ~loc (h, c) }
+    ] ]
+  ;
+  bindings:
+    [ [ test_lpar_idnum_coloneq; bl = LIST1 simple_binding ->
+        { CAst.make ~loc @@ QExplicitBindings bl }
+      | bl = LIST1 Constr.constr ->
+        { CAst.make ~loc @@ QImplicitBindings bl }
+    ] ]
+  ;
+  q_bindings:
+    [ [ bl = bindings -> { bl } ] ]
+  ;
+  q_with_bindings:
+    [ [ bl = with_bindings -> { bl } ] ]
+  ;
+  intropatterns:
+    [ [ l = LIST0 nonsimple_intropattern -> { CAst.make ~loc l } ] ]
+  ;
+(*   ne_intropatterns: *)
+(*     [ [ l = LIST1 nonsimple_intropattern -> l ]] *)
+(*   ; *)
+  or_and_intropattern:
+    [ [ "["; tc = LIST1 intropatterns SEP "|"; "]" -> { CAst.make ~loc @@ QIntroOrPattern tc }
```

Comment:
I noted earlier that these duplicate existing productions in g_tactic.mlg.  Shouldn't this be cleaned up now rather than leaving it as tech debt?  Please comment.

Reply:
Nope, I mentioned that these entries are *not* the same as the Ltac1 ones. They also allow for antiquotations.

Reply:
Sorry I missed your earlier comment.

OK, they do different things.  Are the nonterminals here also completely independent from those with the same name in `g_tactic.mlg`?  Or do the productions for each nonterminal get merged with the definitions in `g_tactic.mlg`?

intropatterns are not just used in Ltac1, for example they're used in a number of tactics.  Does this change the behavior of those tactics?

How long do you expect we'll have both sets of productions?  This seems pretty undesirable and confusing for developers.  Seems like tech debt to me.

Reply:
> intropatterns are not just used in Ltac1, for example they're used in a number of tactics. Does this change the behavior of those tactics?

They're not the same intropatterns, even between Ltac1 and the OCaml implementation. You're confusing parsing with dynamic semantics here, it seems. Ltac2 shares the same dynamic implementation of tactics as Ltac1, but has a different surface language.

> How long do you expect we'll have both sets of productions?

Until we do `git rm plugins/ltac`.

> Seems like tech debt to me.

That's a good way to qualify Ltac.

Reply:
Ah, I don't think you understood me.  Let me make it as concrete as I can:

`g_ltac2.mlg` has a production

```
  or_and_intropattern:
    [ [ "["; tc = LIST1 intropatterns SEP "|"; "]" -> { CAst.make ~loc @@ QIntroOrPattern tc }
```

and `g_tactic.mlg` that have identical syntax but different actions:

```
  or_and_intropattern:
    [ [ "["; tc = LIST1 intropatterns SEP "|"; "]" -> { IntroOrPattern tc }
```

Are these nonterminals combined or somehow kept separate in the parser?  They have the same name.  If they are combined, don't you end up with 2 identical rules for the NT?  Which one wins?
(As I mentioned elsewhere, doc for how the parser combines the various `.mlg` files is not available.)

Also, would you address my earlier question: "intropatterns are not just used in Ltac1, for example they're used in a number of tactics. Does this change the behavior of those tactics?"  When we drop Ltac1, which `intropatterns` will the tactics use?

Reply:
> They have identical syntax but different actions

No, they don't. The `intropatterns` entry is not the same one in each file.

> Are these nonterminals combined or somehow kept separate in the parser?

These are different entries, so they're kept separate. Thankfully, one of the few things that camlp5 got right was that entry naming is merely for user readability. This is not a global table sharing the various entries. (Here you can see that, the types don't even agree.)

> Also, would you address my earlier question: "intropatterns are not just used in Ltac1, for example they're used in a number of tactics. Does this change the behavior of those tactics?" When we drop Ltac1, which intropatterns will the tactics use?

 I actually answered but it seem I was not clear: these are *not* the same intro-patterns, even in Ltac1. There are three different types involved in Ltac1 : raw, glob and top intropatterns. The first one is the syntax, the second one is the compiled form, and the third one is used by the tactic API (i.e., not Ltac per se). Ltac2 reuses the same tactic API, so top intropatterns are shared, but this has nothing to do with syntax, it is OCaml API. When Ltac1 is dropped, we will just remove the raw and glob variants.

Reply:
So what you see on the right of the arrow is different because this is what the syntax is mapped to in the AST, which is separate for Ltac1 and Ltac2, but this doesn't mean that the semantics is different.

Reply:
Clear as mud.  I will want to understand this fully for https://github.com/coq/coq/pull/9884.  I'll take your word on it for now, and we can continue the discussion elsewhere (perhaps in that PR).  But I will say this now:

I'm only asking about parsing, so it seems only "raw" is relevant.  I would like to understand the scoping rules for nonterminals in Campl5.  It sounds like they can have have scope local to the `.mlg` file, perhaps that's the default.  How do you share things across `.mlg` files?  Perhaps by using `GLOBAL`?  I didn't see this info in the Camlp5 doc, which is not a model of clarity.  If it's there, would you point me to it?

While it may useful to have .mlg-local nonterminal/production definitions, it seems to me that it complicates documenting the grammar (2 NTs with the same name but only used in specific contexts) while it also has the potential to confuse developers trying to understand the parsing.  Having some documentation of some of these basics would help immensely-particularly if you may want others learn that code so they might help maintain/improve it.

> So what you see on the right of the arrow is different because this is what the syntax is mapped to in the AST, which is separate for Ltac1 and Ltac2, but this doesn't mean that the semantics is different.

I'm asking only about parsing and which action gets called.  I would certainly hope the semantics are the same.

Reply:
I think it's local to the GRAMMAR EXTEND unless using GLOBAL.

Reply:
> I would like to understand the scoping rules for nonterminals in Campl5

(Global) Camlp5 entries are OCaml values. So they obey the scoping rules of the host language.